raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate.py

```diff
@@ -1814,6 +1814,102 @@ def test_filter_reachable_routes():
     assert possible_routes[1] not in filtered_routes
 
 
+def test_resume_waiting_transfer():
+    """ Test that a mediator who has a waiting_transfer
+    set (the transfer couldn't be sent forward or backward
+    due to availability or capacity or timeout issues) will retry
+    mediating the waiting_transfer as soon as this transfer's
+    lock expiration becomes valid.
+    """
+    setup = factories.make_transfers_pair(2)
+
+    # Also add transfer sender channel
+    partner_state = factories.NettingChannelEndStateProperties(address=UNIT_TRANSFER_SENDER)
+    payer_channel = factories.create(
+        factories.NettingChannelStateProperties(partner_state=partner_state)
+    )
+    setup.channels.channels.append(payer_channel)
+
+    from_hop = factories.make_hop_from_channel(payer_channel)
+    possible_routes = setup.channels.get_routes()
+
+    # we received a transfer that expires after settle timeout would hit
+    lock_expiration = UNIT_SETTLE_TIMEOUT + 3
+    received_transfer = factories.create(
+        factories.LockedTransferSignedStateProperties(
+            amount=1,
+            expiration=lock_expiration,
+            canonical_identifier=payer_channel.canonical_identifier,
+        )
+    )
+
+    # we simulate that the initial transfer came to early (from our point of view)
+    # so it went into `waiting_transfer` state
+    # this can happen, when reveal_timeout is around `settle_timeout / 2` and the
+    # receiving node is slightly behind with the blockchain sync
+    # see: https://github.com/raiden-network/raiden/issues/4998
+    transition_result = mediator.state_transition(
+        mediator_state=None,
+        state_change=ActionInitMediator(
+            from_hop=from_hop,
+            route_states=possible_routes,
+            from_transfer=received_transfer,
+            sender=payer_channel.partner_state.address,
+            balance_proof=received_transfer.balance_proof,
+        ),
+        channelidentifiers_to_channels=setup.channel_map,
+        nodeaddresses_to_networkstates=setup.channels.nodeaddresses_to_networkstates,
+        pseudo_random_generator=random.Random(),
+        block_number=1,
+        block_hash=factories.make_block_hash(),
+    )
+    mediator_state = transition_result.new_state
+    assert mediator_state.waiting_transfer is not None
+
+    too_early = Block(block_number=2, gas_limit=1, block_hash=factories.make_block_hash())
```

Comment:
nitpick: Here and below I would rename the state changes to also have the `_block` suffix in order to show what it is.

So `too_early_block`, `late_enough_block`